Create a way to localize strings

YodaForce157 opened the pull request:
TODO

- [x] Create localized components
- [x] Implement those components
- [x] ~~Move over to Locale Enum?~~
- [x] ~~Find people to do translations/use google translate (its only words so it should be accurate until we get localizers)~~

Commit:
Create sample settings button

Commit:
Silly button won't go where I tell it to :angry:

No Idea why this is, I've removed setAlignment(Pos.CENTER_LEFT) but that didn't seem to do anything

Commit:
Created basic dropdown

Created basic settings dropdown with items. Next: Create enum for each dropdown item, add functionality and then work out how to position correctly

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Add method to delete project dir

Commit:
revert ProjectListCell.java

Ugly dropdown, it worked, but looked terrible and could be implemented better

Commit:
Update ProjectManager.java

Commit:
Made del on a project cell delete project files

Commit:
Focus the window on startup.

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Update to latest PR

Commit:
Update ProjectListCell.java

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Possible way to check if user is shifting?

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Update to current pr

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Add time to log and have only that runtimes logs in the file, also fix appearance pane to be consistent with general pane

Commit:
Add titles to each setting pane

YodaForce157 committed:
Merge branch 'Railroad-Team:main' into settings-menu

YodaForce157 committed:
Merge branch 'Railroad-Team:main' into settings-menu

Commit:
Add localization and keymaps placeholder table

Commit:
Optimize code & create enum

Commit:
Update LocalizationHandler.java

Commit:
Create a backup for if there isn't a key in that language file

Commit:
Add all localizations to en_us.json

Commit:
Move lang files over to .properties

DaRealTurtyWurty commented on src/main/java/io/github/railroad/utility/localization/LocalizationHandler.java:
```diff
@@ -0,0 +1,66 @@
+
+import static io.github.railroad.Railroad.LOGGER;
+
+public class LocalizationHandler {
```
Rename to `I18n`

DaRealTurtyWurty commented on src/main/java/io/github/railroad/utility/localization/LocalizationHandler.java:
```diff
@@ -0,0 +1,66 @@
+        }
+    }
+
+    public static String getLocalized(String key) {
```
Rename to `localize`

DaRealTurtyWurty commented on src/main/java/io/github/railroad/utility/localization/LocalizationHandler.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    public static String getLocalized(String key) {
+        LOGGER.info("Getting localized string for key {}", key);
```
```suggestion
        LOGGER.debug("Getting localized string for key {}", key);
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/utility/localization/LocalizationHandler.java:
```diff
@@ -0,0 +1,66 @@
+            LOGGER.error("Error finding translations for {} {} Moving to english", key, CURRENT_LANG);
+
+            setLanguage(Languages.ENGLISH);
+            getLocalized(key);
```
Also, if its not found in english, this will be infinite recursion. So in that case you should just default to the `key` 
```suggestion
            return getLocalized(key);
```

DaRealTurtyWurty commented on src/main/resources/io/github/railroad/lang/en_us.properties:
```diff
@@ -0,0 +1,41 @@
+railroad.home.settings.keymaps.action= Action
+
+railroad.home.settings.keymaps.action.runproject= Run Project
+railroad.home.settings.keymaps.action.termproject= Terminate Project
```
Any reason for the spaces after `=`? Looks really weird

DaRealTurtyWurty commented on src/main/java/io/github/railroad/settings/ui/general/SettingsGeneralPane.java:
```diff
@@ -21,7 +29,23 @@ public SettingsGeneralPane() {
+        langTitleLabel.setStyle("-fx-font-weight: bold;");
+
+        var langComboBox = new ComboBox<>();
+        langComboBox.getItems().addAll("English", "Español", "Français", "Deutsch");
```
This should use the enum.

DaRealTurtyWurty commented on src/main/java/io/github/railroad/settings/ui/general/SettingsGeneralPane.java:
```diff
@@ -21,7 +29,23 @@ public SettingsGeneralPane() {
+        langComboBox.setOnAction(e -> {
+            LocalizationHandler.setLanguage(Languages.fromLanguage(langComboBox.getValue().toString()));
+            //TODO reload the UI
+            Railroad.getWindow().close();
```
No need for this

DaRealTurtyWurty commented on src/main/java/io/github/railroad/utility/localization/Languages.java:
```diff
@@ -0,0 +1,44 @@
+
+        return null;
+    }
+}
```
We could possibly consider just using `Locale` instead of this.

YodaForce157 replied:
I'm going to keep it using Languages for this PR, or atleast at the moment, I might move it over later but I'll prioritize updating all the components first.

YodaForce157 committed:
Merge branch 'Railroad-Team:main' into localizer

Commit:
Move files around

Commit:
DO NOT PUSH

Commit:
Fix text not showing

Commit:
Correctly bind labels

Commit:
Fix bugs and create LocalizedComponents

Commit:
add translations

CHATGPT GENERATED TRANSLATIONS

Commit:
move to localized

DaRealTurtyWurty committed:
Fix localization

Commit:
move to localized

Commit:
changes I think

Commit:
More moving

Commit:
Finish up

Commit:
Merge branch 'heads/upstream/main' into localizer

Commit:
fix imports

YodaForce157 commented:
Ignore the conflict (I think) as I can't seem to do anything on my end.

Commit:
Merge branch 'temp' into localizer

Commit:
git can be so useless

Commit:
e

Commit:
Update GithubAccounts.java

YodaForce157 committed:
Merge branch 'Railroad-Team:main' into temp

Commit:
Merge branch 'temp' into localizer

Commit:
Pray that this is fine.

I've ported over my previous code to this, however there are still a few components that need localizing. I will do that ASAP. I am also waiting on the form builder to port over that.

Commit:
Merge branch 'localizer' into localizernew
Move over code to original PR

YodaForce157 commented:
Should be OK now. I'll wait until this is merged/commented on to touch anything else because I'd rather not break everything _again_

Once merged I'll create another pr to finish up the few leftover components that haven't been localized and also finish up the language files.

Commit:
Fix themes showing incorrect colour for LocalizedText

DaRealTurtyWurty commented on src/main/java/io/github/railroad/Railroad.java:
```diff
@@ -32,6 +35,8 @@
 import java.util.function.Consumer;
 
+import static io.github.railroad.utility.localization.L18n.localize;
+
```
```suggestion
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/Railroad.java:
```diff
@@ -151,7 +157,7 @@ public void start(Stage primaryStage) {
         primaryStage.setMinHeight(scene.getHeight() + 10);
         primaryStage.setScene(scene);
-        primaryStage.setTitle("Railroad - 1.0.0(dev)");
+        primaryStage.setTitle(localize("railroad.home.window.title") + " - 1.0.0(dev)");
```
The title of the application should not be localized. 
```suggestion
        primaryStage.setTitle(localize("Railroad - 1.0.0(dev)");
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/project/ProjectSort.java:
```diff
@@ -5,23 +5,23 @@
     public String getName() {
-        return name;
+        return key;
     }
```
If this returns the key then the method should be called `getKey`. If it should still return the name then it should be localized.

DaRealTurtyWurty commented on src/main/java/io/github/railroad/project/ui/create/widget/ProjectSortComboBox.java:
```diff
@@ -1,42 +1,41 @@
+        });
+
         setValue(ProjectSort.NONE);
     }
```
Any reason you can't use a `LocalizedListCell`?

YodaForce157 replied:
Because that isn't a dropdown? Unless I'm missing something. Changing it to a localized list cell just removes the dropdown part of it and messes with theming.

DaRealTurtyWurty replied:
We're doing `setButtonCell(new ListCell<>() {`

DaRealTurtyWurty commented on src/main/java/io/github/railroad/settings/SettingsCategory.java:
```diff
@@ -6,30 +6,30 @@
     public String getName() {
-        return name;
+        return key;
     }
```
If this returns the key then the method should be called `getKey`. If it should still return the name then it should be localized.

DaRealTurtyWurty commented on src/main/java/io/github/railroad/settings/ui/SettingsBehaviorPane.java:
```diff
@@ -1,6 +1,22 @@
-public class SettingsBehaviorPane extends BorderPane {
-}
+public class SettingsBehaviorPane extends RRVBox {
+
```
```suggestion
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/settings/ui/SettingsKeymapsPane.java:
```diff
@@ -1,6 +1,38 @@
-public class SettingsKeymapsPane extends BorderPane {
-}
+public class SettingsKeymapsPane extends RRVBox {
+
```
```suggestion
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/settings/ui/SettingsProjectsPane.java:
```diff
@@ -1,6 +1,21 @@
-public class SettingsProjectsPane extends BorderPane {
-}
+public class SettingsProjectsPane extends RRVBox {
+
```
```suggestion
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/settings/ui/SettingsToolsPane.java:
```diff
@@ -1,6 +1,21 @@
-public class SettingsToolsPane extends BorderPane {
-}
+public class SettingsToolsPane extends RRVBox {
+
```
```suggestion
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/settings/ui/general/SettingsGeneralPane.java:
```diff
@@ -1,18 +1,38 @@
         titleLabel.setAlignment(Pos.CENTER);
 
-        getChildren().addAll(titleLabel);
+        var langBox = new VBox(10);
```
```suggestion
        var langBox = new RRVBox(10);
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/settings/ui/general/SettingsGeneralPane.java:
```diff
@@ -1,18 +1,38 @@
 
+import java.util.Arrays;
+
 public class SettingsGeneralPane extends VBox {
```
```suggestion
public class SettingsGeneralPane extends RRVBox {
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/settings/ui/general/SettingsGeneralPane.java:
```diff
@@ -1,18 +1,38 @@
 import javafx.geometry.Pos;
-import javafx.scene.control.Label;
+import javafx.scene.control.ComboBox;
 import javafx.scene.layout.VBox;
```
```suggestion
import io.github.railroad.ui.defaults.RRVBox;
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/ui/localized/LocalizedComboBox.java:
```diff
@@ -0,0 +1,30 @@
+            public String toString(T object) {
+                if (object == null) {
+                    return "NullObject";
+                }
```
```suggestion
                if (object == null)
                    return "NullObject";

```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/ui/localized/LocalizedComboBox.java:
```diff
@@ -0,0 +1,30 @@
+                if (key == null) {
+                    return "NullKey";
+                }
+                return L18n.localize(key);
```
```suggestion
                if (key == null)
                    return "NullKey";

                return L18n.localize(key);
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/ui/localized/LocalizedComboBox.java:
```diff
@@ -0,0 +1,30 @@
+                }
+                return L18n.localize(key);
+            }
+            @Override
```
```suggestion
            }
            
            @Override
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/ui/localized/LocalizedLabel.java:
```diff
@@ -0,0 +1,34 @@
+
+public class LocalizedLabel extends Label {
+    private String currentKey;
+    AtomicReference<String> argString = new AtomicReference<>("");
```
```suggestion
    private final AtomicReference<String> argString = new AtomicReference<>("");
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/ui/localized/LocalizedTooltip.java:
```diff
@@ -0,0 +1,34 @@
+
+public class LocalizedTooltip extends Tooltip {
+    private String currentKey;
+    AtomicReference<String> argString = new AtomicReference<>("");
```
```suggestion
    private final AtomicReference<String> argString = new AtomicReference<>("");
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/utility/localization/L18n.java:
```diff
@@ -0,0 +1,69 @@
+        Languages newLang = Languages.valueOf(ConfigHandler.getConfigJson().get("settings").getAsJsonObject().get("language").getAsString());
+
+        try {
+            InputStream props = Railroad.getResourceAsStream("lang/" + newLang + ".properties");
```
Before we go too far with this, its probably worth having them be `.lang` extensions, not `.properties`

DaRealTurtyWurty commented on src/main/java/io/github/railroad/utility/localization/L18n.java:
```diff
@@ -0,0 +1,69 @@
+            setLanguage(Languages.en_us);
+            return localize(key);
+        }
+        return LANG_CACHE.get(key).toString();
```
```suggestion
        }
        
        return LANG_CACHE.get(key).toString();
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/utility/localization/Languages.java:
```diff
@@ -0,0 +1,32 @@
+    en_us("English"),
+    es_es("Español"),
+    fr_fr("Français"),
+    de_de("Deutsch");
```
```suggestion
    EN_US("English"),
    ES_ES("Español"),
    FR_FR("Français"),
    DE_DE("Deutsch");
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/utility/localization/Languages.java:
```diff
@@ -0,0 +1,32 @@
+        }
+        LOGGER.error("Cannot find language: {}", name);
+        return null;
+    }
```
```suggestion
    public static Languages fromName(String name) {
        for(Languages lang : Languages.values()) {
            if(lang.name().equalsIgnoreCase(name)) {
                return lang;
            }
        }
        
        LOGGER.error("Cannot find language: {}", name);
        return null;
    }
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/utility/localization/Languages.java:
```diff
@@ -0,0 +1,32 @@
+
+import static io.github.railroad.Railroad.LOGGER;
+
+public enum Languages {
```
Rename to `Language`

DaRealTurtyWurty commented on src/main/java/io/github/railroad/welcome/WelcomeLeftPane.java:
```diff
@@ -35,7 +37,7 @@ public WelcomeLeftPane() {
         var rightVbox = new RRVBox();
         rightVbox.setAlignment(Pos.CENTER);
-        rightVbox.getChildren().add(new Label("Railroad IDE"));
+        rightVbox.getChildren().add(new LocalizedLabel("railroad.home.window.title"));
```
Window title should not be localized
```suggestion
        rightVbox.getChildren().add(new Label("Railroad IDE"));
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/welcome/WelcomeLeftPane.java:
```diff
@@ -65,28 +67,28 @@ public RRListView<MenuType> getListView() {
         public String getName() {
-            return this.name;
+            return this.key;
         }
```
If this returns the key then the method should be called `getKey`. If it should still return the name then it should be localized.

DaRealTurtyWurty commented on themes/cupertino-dark.css:
```diff
@@ -136,6 +136,10 @@ Text {
 
+LocalizedText {
+  -fx-fill: -color-fg-default;
+}
```
You only need to do this in base.css. Not in all the themes.

Commit:
Requested Changes

Commit:
Update button cell to LocalizedListCell

YodaForce157 committed:
Merge branch 'Railroad-Team:main' into localizer

Commit:
minor fix

DaRealTurtyWurty commented on src/main/resources/io/github/railroad/styles/default-light.css:
```diff
@@ -132,7 +132,7 @@
-Text {
+LocalizedText {
     -fx-fill: -color-fg-default;
 }
```
```suggestion
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/utility/localization/L18n.java:
```diff
@@ -0,0 +1,70 @@
+    public static void loadLanguage() {
+        //Loads the language into cache and sets the CURRENT_LANG
+        LOGGER.info("Loading language file");
+        Language newLang = Language.valueOf(ConfigHandler.getConfigJson().get("settings").getAsJsonObject().get("language").getAsString());
```
```suggestion
        var newLang = Language.valueOf(ConfigHandler.getConfigJson().get("settings").getAsJsonObject().get("language").getAsString());
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/utility/localization/L18n.java:
```diff
@@ -0,0 +1,70 @@
+        //Updates the config and calls loadLanguage to update the cache and CURRENT_LANG
+        LOGGER.debug("Setting language to {}", language);
+
+        var updated = ConfigHandler.getConfigJson();
```
```suggestion
        JsonObject updated = ConfigHandler.getConfigJson();
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/settings/ui/SettingsKeymapsPane.java:
```diff
@@ -1,6 +1,37 @@
+        //TODO add items to table from config
+        var keymapsList = new TableView<>();
+        TableColumn keyColumn = new TableColumn("Key");
+        TableColumn actionColumn = new TableColumn("Action");
```
```suggestion
        var keyColumn = new TableColumn("Key");
        var actionColumn = new TableColumn("Action");
```

DaRealTurtyWurty commented on src/main/java/io/github/railroad/settings/ui/SettingsBehaviorPane.java:
```diff
@@ -1,6 +1,21 @@
-}
+public class SettingsBehaviorPane extends RRVBox {
+    public SettingsBehaviorPane() {
+        Label title = new LocalizedLabel("railroad.home.settings.behavior");
```
```suggestion
        var title = new LocalizedLabel("railroad.home.settings.behavior");
```

Commit:
Requested changes

Commit:
Update en_us.lang

Commit:
Remove unnecessary bit

Commit:
Create LocalizedTextField component and update language files

DaRealTurtyWurty commented on src/main/java/io/github/railroad/project/ui/create/widget/ProjectSearchField.java:
```diff
@@ -1,9 +1,9 @@
-        setPromptText("Search projects...");
+        super("railroad.home.welcome.projectsearch");
     }
 }
```
Since you're touching this, you might as well yeet this custom class for it. We don't need it to be it's own class

Commit:
Remove ProjectSearchField

Commit:
Fix langs with special characters

Commit:
fix startup issue
